### PR TITLE
Stricter Skin Definitions

### PR DESCRIPTION
**Pull request by marstaik**
There are some common ambiguities with the current skin definition that leave a lot of implementation details up in the air, for reference see here: https://github.com/KhronosGroup/glTF/issues/1665

This is an initial PR to get some feedback as to how to potentially tighten up the Skinning specifications a bit and remove unnecessary ambiguity.

### CODE REVIEW

**Comment by marstaik** [comment]
I have devised a way that should not break any of the current importers, and only benefit those engines trying to import with game engines, if you would still be interested in an updated PR.

**Comment by donmccurdy** [comment]
@marstaik suggestions for a non-breaking alternative would be welcome, yes.

I would further clarify @emackey's comment, though, by saying that not breaking existing importers alone is not sufficient for a 2.x change – we also need to avoid invalidating existing glTF 2.0 models, unless they're egregiously unsupportable in some way. That would require a longer-term v3.0, or perhaps an extension (`EXT_skinning_strict`?) that imposes additional requirements on the core specification.

File: specification/2.0/README.md [hunk 1]
```diff
@@ -862,7 +863,13 @@ After applying morph targets to vertex positions and normals, tangent space may
...
+All skins are stored in the `skins` array of the asset. Each skin is defined by the `inverseBindMatrices` property (which points to an accessor with IBM data), used to bring coordinates being skinned into the same space as each joint; and a `joints` array property that lists the nodes indices used as joints to animate the skin. The order of joints is defined in the `skin.joints` array and it must match the order of `inverseBindMatrices` data.
+
+All joints defined in a skin must form a connected subtree within the node hierarchy, and the highest order node is the `skin parent`. If a skin contains multiple distinct subtrees (multi-rooted, a joint is not defined which connects them) then the "roots" of the distict subtrees must be siblings in the node hierarchy, and their comment parent is the `skin parent` (note that in this case, the `skin parent` may not be a joint).
```
Comment (donmccurdy) [inline specification/2.0/README.md]: Given the motivation for these restrictions, why allow multiple distinct subtrees at all? Unless I've missed something, couldn't any skin with multiple distinct subtrees and a common parent be made into a single connected subtree, by turning that common parent into a joint?

**Comment by marstaik** [comment]
While I would welcome having them under a single parent joint, the multiple subtrees (that are siblings) don't pose a problem, as you can group them under the parent inside another entity (skeleton). I feel like whether or not you want to allow multiple joint subtrees is a smaller issue than the whole non-joints between skins issue.

In https://github.com/godotengine/godot/pull/32213 I successfully did some work interpolating overlapping/touching skins to a single skeleton in godot, but that code is very algorithmic in nature, and it seems there are some few edge cases still.

In that PR, you will see joint "stealing" and fake joint creation, expansion of skins to catch non-joints and create fake joints, as well as expanding skins vertically until they find a common parent, then merging skin subsets into skeletons. Before the scene based skin support there was also code to reroot the skin IBMs to the interpolated skeleton.

The point in trying to make is that while it's possible to handle this through code (everything is possible) it's still a huge challenge.

The exporter could make these decisions far for easily and also make sure it's making the correct decisions for the export due to every programs internal modeling differences.

But the PR would not have achieved as much without @reduz implementing Skins in godot.

**Comment by marstaik** [comment]
Sorry that I took so long to respond, been busy.

No, there isn't a strong motivation for that exception. Some systems allow for multirooted trees, so I thought of also allowing it.

However, the strong counter argument is that if a system can handle a multirooted tree, it can definitely handle a single rooted tree.
But the reverse is not true.

So it may just be better to remove allowing multirooted trees.

But, after experiencing first hand the behavior of many importers and exporters, and analyzing gltf behavior, I will try to make a new PR thst captures the real essence of the problems, and tries to correct them in an intelligent way.

**Comment by donmccurdy** [comment]
Closing this PR in favor of #1695 or #1747. We can't make breaking changes to the glTF 2.0 specification at this point, but an extension or changes in the next version of glTF (whenever that might be) are things we can consider.

**Pull request closed without merging**